src/Sparrow/Collections/LockFreeRingBuffer.cs

```diff
@@ -0,0 +1,172 @@
+﻿using System;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Threading;
+using Sparrow.Binary;
+
+namespace Sparrow.Collections
+{
+    public sealed class LockFreeRingBuffer<T>
+    {
+        // Explicit layout to control the exact memory layout
+        [StructLayout(LayoutKind.Sequential, Size = 128)]
+        internal unsafe struct Cell
+        {
+            // Sequence number at offset 0
+            internal long _sequence;
+
+            private fixed byte _sequencePadding[64 - sizeof(long)];
+
+            // Value at offset 64 to ensure it's on a different cache line
+            internal T _value;
+
+            public long Sequence
+            {
+                get => Volatile.Read(ref _sequence);
+                set => Volatile.Write(ref _sequence, value);
+            }
```

Comment:
I think that hiding those in the property is confusing. To properly understand, you needed to add the comments showing this.

Reply:
I agree. It doesn't say anything about allowed read / write usages.